pkg/apis/events/validation/validation.go

```diff
@@ -0,0 +1,96 @@
+func ValidateEvent(event *core.Event) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// Because go
+	zeroTime := time.Time{}
+
+	if event.EventTime.Time == zeroTime {
+		allErrs = append(allErrs, field.Required(field.NewPath("eventTime"), ""))
+	}
+	if len(event.InvolvedObject.Namespace) == 0 && event.Namespace != metav1.NamespaceSystem {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "does not match event.namespace"))
+	}
+	if len(event.ReportingController) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reportingController"), ""))
+	}
+	for _, msg := range validation.IsQualifiedName(event.ReportingController) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reportingController"), event.ReportingController, msg))
+	}
+	if len(event.ReportingInstance) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reportingInstance"), ""))
+	}
+	if len(event.ReportingInstance) > reportingInstanceLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reportingInstance"), "", fmt.Sprintf("can have at most %v characters", reportingInstanceLengthLimit)))
+	}
+	if len(event.Action) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("action"), ""))
+	}
+	if len(event.Action) > actionLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("action"), "", fmt.Sprintf("can have at most %v characters", actionLengthLimit)))
+	}
+	if len(event.Reason) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reason"), ""))
+	}
+	if len(event.Reason) > reasonLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reason"), "", fmt.Sprintf("can have at most %v characters", reasonLengthLimit)))
+	}
+	if len(event.Message) > noteLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("message"), "", fmt.Sprintf("can have at most %v characters", noteLengthLimit)))
+	}
+	if event.Type != v1.EventTypeNormal && event.Type != v1.EventTypeWarning {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("type"), "", fmt.Sprintf("has invalid value: %v", event.Type)))
+	}
+	if event.FirstTimestamp.Time != zeroTime {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("firstTimestamp"), "", "needs to be unset"))
+	}
+	if event.LastTimestamp.Time != zeroTime {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("lastTimestamp"), "", "needs to be unset"))
+	}
+	if event.Count != 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("count"), "", "needs to be unset"))
+	}
+	if event.Source.Component != "" || event.Source.Host != "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("source"), "", "needs to be unset"))
+	}
+
+	for _, msg := range validation.IsDNS1123Subdomain(event.Namespace) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("namespace"), event.Namespace, msg))
+	}
+	return allErrs
+}
```

Comment:
Can you please also:
- squash all non-autogenerated commits to a single one
- squash all autogenerated commits to another (but single) one
?